### PR TITLE
Add minimal support for missing routes in alertmanager distributor.

### PR DESCRIPTION
**Pull request by stevesg**
**What this PR does**:
Add and test the following routes to the alertmanager distributor (in use only when sharding is enabled):
- `POST   /alertmanager/api/v1/silences`
- `GET    /alertmanager/api/v1/silences`
- `GET    /alertmanager/api/v1/silence/{id}`
- `DELETE /alertmanager/api/v1/silence/{id}`
- `GET    /alertmanager/api/v1/status`
- `POST   /alertmanager/api/v1/receivers`

All requests are simply forwarded to a single randomly chosen replica
of the shard which contains the user.

~~Note I've had to incorporate the fix from https://github.com/cortexproject/cortex/pull/4083 in this PR.~~ Rebased.

**Checklist**
- [X] Tests updated
- [X] ~~Documentation added~~
- [X] ~~`CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`~~

### CODE REVIEW

**Comment by stevesg** [comment]
Draft while I fix the tests.

File: integration/e2ecortex/client.go [hunk 1]
```diff
@@ -509,15 +509,127 @@ func (c *Client) SendAlertToAlermanager(ctx context.Context, alert *model.Alert)
 	return nil
 }
 
-func (c *Client) CreateSilence(ctx context.Context, silence types.Silence) error {
+func (c *Client) CreateSilence(ctx context.Context, silence types.Silence) (string, error) {
```
Comment (pstibrany) [inline integration/e2ecortex/client.go:512]: Comment about what string return value is would be nice.

**Review by pstibrany — APPROVED** [review]
LGTM, nice tests! Thanks.

File: pkg/alertmanager/distributor.go [hunk 2]
```diff
@@ -86,11 +107,30 @@ func (d *Distributor) DistributeRequest(w http.ResponseWriter, r *http.Request)
 
 	logger := util_log.WithContext(r.Context(), d.logger)
 
+	if r.Method == http.MethodPost {
+		if d.isQuorumWritePath(r.URL.Path) {
+			d.doWrite(userID, w, r, logger)
```
Comment (pracucci) [inline pkg/alertmanager/distributor.go:112]: Should we rename `doWrite()` to `doQuorumWrite()`?

**Review by pracucci — APPROVED** [review]
Solid work, congrats! No comments, LGTM! 👏

**Pull request merged**